src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -74,27 +78,80 @@ protected override void OnEventWritten(EventWrittenEventArgs eventData)
             return;
         }
 
+        var eventSourceName = eventData.EventSource.Name;
+
+        if (eventData.EventName != "EventCounters")
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreNonEventCountersName(eventSourceName);
+            return;
+        }
+
+        if (eventData.Payload.Count == 0 || eventData.Payload[0] is not IDictionary<string, object>)
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenEventArgsPayloadNotParseable(eventSourceName);
+            return;
+        }
+
         var payload = eventData.Payload[0] as IDictionary<string, object>;
-        var name = payload["Name"].ToString();
-        var isGauge = payload.ContainsKey("Mean");
-        Tuple<string, string> metricKey = new(eventData.EventSource.Name, name);
+        var hasName = payload.TryGetValue("Name", out var nameObj);
 
-        if (isGauge)
+        if (!hasName)
         {
-            this.values[metricKey] = Convert.ToDouble(payload["Mean"]);
+            EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenEventArgsWithoutName(eventSourceName);
+            return;
         }
-        else
+
+        var name = nameObj.ToString();
+
+        var hasMean = payload.TryGetValue("Mean", out var mean);
+        var hasIncrement = payload.TryGetValue("Increment", out var increment);
+
+        if (!(hasIncrement ^ hasMean))
         {
-            _ = this.values.TryAdd(metricKey, 0);
-            this.values[metricKey] += Convert.ToDouble(payload["Increment"]);
+            EventCountersInstrumentationEventSource.Log.IgnoreMeanIncrementConflict(eventSourceName);
+            return;
         }
 
-        var instrumentName = $"{eventData.EventSource.Name}.{name}";
-        _ = this.instruments.TryAdd(
-            metricKey,
-            isGauge ? MeterInstance.CreateObservableGauge(instrumentName, () => this.values[metricKey]) : MeterInstance.CreateObservableCounter(instrumentName, () => this.values[metricKey]));
+        var value = Convert.ToDouble(hasMean ? mean : increment);
+        this.EventWritten(hasMean, eventSourceName, name, value);
     }
 
     private static Dictionary<string, string> GetEnableEventsArguments(EventCountersInstrumentationOptions options) =>
         new() { { "EventCounterIntervalSec", options.RefreshIntervalSecs.ToString() } };
+
+    private static bool IsValidInstrumentName(string name) =>
+        !string.IsNullOrWhiteSpace(name) && InstrumentNameRegex.IsMatch(name);
+
+    private void EventWritten(bool isGauge, string eventSourceName, string name, double value)
```

Comment:
Could you rename this method because it hides the inherited member 'EventWritten'.